datahub-graphql-api/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java

```diff
@@ -0,0 +1,157 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static graphql.schema.idl.RuntimeWiring.newRuntimeWiring;
+
+/**
+ * Friendly wrapper over a {@link GraphQL} instance.
+ */
+public class GraphQLEngine {
+
+    private final GraphQL _engine;
+    private final Map<String, Supplier<DataLoader<?, ?>>> _dataLoaderSuppliers;
+
+    private GraphQLEngine(@Nonnull final List<String> schemas,
+                          @Nonnull final RuntimeWiring runtimeWiring,
+                          @Nonnull final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers) {
+
+        _dataLoaderSuppliers = dataLoaderSuppliers;
+
+        /*
+         * Parse schema
+         */
+        SchemaParser schemaParser = new SchemaParser();
+        TypeDefinitionRegistry typeDefinitionRegistry = new TypeDefinitionRegistry();
+        schemas.forEach(schema -> typeDefinitionRegistry.merge(schemaParser.parse(schema)));
+
+        /*
+         * Configure resolvers (data fetchers)
+         */
+        SchemaGenerator schemaGenerator = new SchemaGenerator();
+        GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);
+
+        /*
+         * Instantiate engine
+         */
+        _engine = GraphQL.newGraphQL(graphQLSchema).build();
+    }
+
+    public ExecutionResult execute(@Nonnull final String query,
+                                   @Nullable final Map<String, Object> variables,
+                                   @Nonnull final QueryContext context) {
+        /*
+         * Init DataLoaderRegistry - should be created for each request.
+         */
+        DataLoaderRegistry register = createDataLoaderRegistry(_dataLoaderSuppliers);
+
+        /*
+         * Construct execution input
+         */
+        ExecutionInput executionInput = ExecutionInput.newExecutionInput()
+            .query(query)
+            .variables(variables)
+            .dataLoaderRegistry(register)
+            .context(context)
+            .build();
+
+        /*
+         * Execute GraphQL Query
+         */
+        return _engine.execute(executionInput);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Used to construct a {@link GraphQLEngine}.
+     */
+    public static class Builder {
+
+        private final List<String> _schemas = new ArrayList<>();
+        private final Map<String, Supplier<DataLoader<?, ?>>> _loaderSuppliers = new HashMap<>();
+        private final RuntimeWiring.Builder _runtimeWiringBuilder = newRuntimeWiring();
+
+        /**
+         * Used to add a schema file containing the GQL types resolved by the engine.
+         *
+         * If multiple files are provided, their schemas will be merged together.
+         */
+        public Builder addSchema(final String schema) {
```

Comment:
can this ever throw exceptions?

Reply:
Runtime exceptions can be thrown at schema parsing time (at build) within the GraphQL-provided schema parser